indico/web/client/js/custom_elements/ind_share_widget.js

```diff
@@ -0,0 +1,392 @@
+          aria-modal
+        />
+      }
+      content={
+        <>
+          <Header
+            size="huge"
+            icon="share alternate"
+            content={Translate.string('Share this page')}
+            styleName="header-text"
+          />
+          <Header size="tiny" styleName="share-section-header">
+            <Icon name="linkify" styleName="icon" />
+            <HeaderContent styleName="title">
+              <Translate>Direct link</Translate>
+            </HeaderContent>
+          </Header>
+          <CopyToClipboard eventUrl={eventUrl} />
+          <Header size="tiny" styleName="share-section-header">
+            <Icon name="calendar alternate outline" styleName="icon" />
+            <HeaderContent styleName="title">
+              <Translate>Add to calendar</Translate>
+            </HeaderContent>
+          </Header>
+          <CalendarButtons googleCalParams={googleCalParams} outlookCalParams={outlookCalParams} />
+          <Header size="tiny" styleName="share-section-header">
+            <Icon name="share square" styleName="icon" />
+            <HeaderContent styleName="title">
+              <Translate>Share on social media</Translate>
+            </HeaderContent>
+          </Header>
+          <Grid columns={2} stretched>
+            {!_.isEmpty(Indico.User) ? (
+              <SocialButton
+                name="Mastodon"
+                logo={`${Indico.Urls.ImagesBase}/mastodon.svg`}
+                color="violet"
+                render={button => (
+                  <SetupMastodonServer
+                    button={button}
+                    setMastodonOpen={setMastodonOpen}
+                    urlParams={shareUrlText(eventName, eventStartDt, eventUrl)}
+                  />
+                )}
+                button
+                buttonProps={{
+                  as: 'a',
+                  target: '_blank',
+                  href: Indico.User.mastodonServerURL
+                    ? `${Indico.User.mastodonServerURL}/share?text=${eventName} (${eventStartDt}) · Indico (${eventUrl})`
+                    : null,
+                }}
+              />
+            ) : (
+              <SocialButton
+                name="Mastodon"
+                logo={`${Indico.Urls.ImagesBase}/mastodon.svg`}
+                color="grey"
+                render={button => (
+                  <Popup
+                    trigger={button}
+                    content={Translate.string('Please login to share on Mastodon')}
+                    position="top right"
+                    wide
+                  />
+                )}
+                button
+              />
+            )}
+            <SocialButton
+              name="Twitter"
+              logo={`${Indico.Urls.ImagesBase}/twitter.svg`}
+              url={`https://twitter.com/intent/tweet?text=${eventName} (${eventStartDt}) · Indico (${eventUrl})`}
+              color="blue"
+              render={button => button}
+            />
```

Comment:
At the risk of being too nitpicky, but I feel the `SocialButton` component is pretty over-engineered just so that it can abstract two buttons. You can somewhat tell from the complexity of the parameters that need to be passed. I think splitting it into two components `TwitterButton` and `MastodonButton` might be better and reduce the complexity a bit.

Here's what I have in mind:

```jsx
function TwitterButton({url}) {
  return (
    <GridColumn styleName="share-button-column">
      <Button as="a" href={url} target="_blank" basic color="blue" styleName="share-button">
        <img src={`${Indico.Urls.ImagesBase}/twitter.svg`} alt="Twitter" />
        <span>Twitter</span>
      </Button>
    </GridColumn>
  );
}

function MastodonButton({urlParams, hasUser, setMastodonOpen}) {
  const href = Indico.User.mastodonServerURL
    ? `${Indico.User.mastodonServerURL}/share?text=${urlParams}`
    : null;

  const button = (
    <Button
      basic
      color={hasUser ? 'violet' : 'grey'}
      styleName="share-button"
      as="a"
      target="_blank"
      href={href}
    >
      <img src={`${Indico.Urls.ImagesBase}/mastodon.svg`} alt="Mastodon" />
      <span>Mastodon</span>
    </Button>
  );

  return (
    <GridColumn styleName="share-button-column">
      {hasUser ? (
        <SetupMastodonServer
          button={button}
          setMastodonOpen={setMastodonOpen}
          urlParams={urlParams}
        />
      ) : (
        <Popup
          trigger={button}
          content={Translate.string('Please login to share on Mastodon')}
          position="top right"
          wide
        />
      )}
    </GridColumn>
  );
}

// to render them:
<Grid columns={2} stretched>
  {!_.isEmpty(Indico.User) ? ( // This check could also go inside the component
    <MastodonButton
      hasUser
      urlParams={shareUrlText(eventName, eventStartDt, eventUrl)}
      setMastodonOpen={setMastodonOpen}
    />
  ) : (
    <MastodonButton urlParams={shareUrlText(eventName, eventStartDt, eventUrl)} />
  )}
  <TwitterButton
    url={`https://twitter.com/intent/tweet?text=${shareUrlText(
      eventName,
      eventStartDt,
      eventUrl
    )}`}
  />
</Grid>
```

Pardon the lack of i18n ;)
I feel like this is overall easier to digest and it keeps the calling code simpler while the logic is nicely encapsulated in the respective components. What do you think?

Reply:
Yeah, this seems much better to digest. I'll give this a shot 🚀